### PR TITLE
Tidy up and document Rust SDK

### PR DESCRIPTION
**Pull request by daviddrysdale**
The docs/sdk.md file in the final commit may help to give the overall shape of this PR, but changes might be easier to review commit-by-commit.

### CODE REVIEW

File: rust/oak/src/grpc/mod.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    /// Process a single gRPC method invocation.
+    ///
+    /// The method name is provided by `method` and the incoming serialized gRPC
+    /// request is held in `req`.  Response messages should be written to `out`,
+    /// as serialized [`GrpcResponse`] messages encapsulating the service
+    /// response.
+    ///
+    /// [`GrpcResponse`]: crate::proto::grpc_encap::GrpcResponse
+    fn invoke(&mut self, method: &str, req: &[u8], out: WriteHandle);
```
Comment (tiziano88) [inline rust/oak/src/grpc/mod.rs]: Could `out` encapsulate `ResponseWriter` directly?

File: rust/oak/src/io/mod.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+//
+
+//! Wrappers for Oak SDK types to allow their use with std::io.
+
+use crate::{channel_write, OakStatus};
+use std::io;
+
+/// Wrapper for a WriteHandle to implement the [`std::io::Write`] trait.
+///
+/// Methods taking `Write` trait objects require mutable references,
```
Comment (tiziano88) [inline rust/oak/src/io/mod.rs]: Not sure I understand why `WriteHandle` is not sufficient?

File: rust/oak/src/io/mod.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        result_from_status(Some(status), buf.len())
+    }
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+/// Map an [`OakStatus`] value to the nearest available [`std::io::Result`].
+fn result_from_status<T>(status: Option<OakStatus>, val: T) -> std::io::Result<T> {
+    match status {
```
Comment (tiziano88) [inline rust/oak/src/io/mod.rs]: Could transform this into a two-level match if you want:

```rust
use OakStatus::*;
match status {
  Some(status) => match status {
    OAK_STATUS_UNSPECIFIED => ...
    OK => ...
  }
  None => ...
}
```

File: rust/oak/src/io/mod.rs (hunk 2 above)
Comment (daviddrysdale) [reply rust/oak/src/io/mod.rs]: It just felt &hellip; untidy.  The `Handle` types are `Copy` / POD types and so generally passed by value, whereas things using `Write` generally want a mutable reference; also, pulling the wrapper into an `io` submodule meant that all of the `std:io` stuff is in one place, and `oak/lib.rs` entirely deals with `OakStatus` values.  Happy to drop the `io` submodule if you feel strongly about it though.

**Pull request merged**